# Jupyter notebook_Elias Franklin da Silva Guedes Tardim
#### Este projeto pretende realizar uma análise do banco de dados "default_of_credit_card_clients__courseware_version_1_21_19.xls", que contêm informações sobre as linhas de crédito de uma "empresa", onde ficam presentes os dados de pagamento e se o cliente se tornou, ou não, um inadimplente. Seguindo o seguinte pretexto:  

#### """**O problema da empresa**

#### Nosso cliente é uma empresa de cartão de crédito. Eles nos trouxeram um dataset que inclui dados demográficos e dados financeiros recentes (últimos seis meses) de uma amostra de 30.000 titulares de contas. Esses dados estão no nível de conta de crédito; em outras palavras, há uma linha para cada conta (você deve sempre esclarecer qual é a definição de linha, em um dataset). As linhas são rotuladas de acordo com se no mês seguinte ao período de dados histórico de seis meses um proprietário de conta ficou inadimplente, ou seja, não fez o pagamento mínimo."""

In [1]:
import pandas as pd
import openpyxl
import numpy as np

df = pd.read_excel('ARQUIVOS/default_of_credit_card_clients__courseware_version_1_21_19.xls')
df_limpa = df.copy()
df_limpa['AGE'] = df_limpa['AGE'].replace(0, np.nan)
df_limpa.dropna(inplace=True)

df_limpa.head(50)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE   AGE          PAY_1  \
0   798fc410-45c1      20000    2          2         1  24.0              2   
1   8a8c8f3b-8eb4     120000    2          2         2  26.0             -1   
2   85698822-43f5      90000    2          2         2  34.0              0   
3   0737c11b-be42      50000    2          2         1  37.0              0   
4   3b7f77cc-dbc0      50000    1          2         1  57.0             -1   
5   ceae2946-47e4      50000    1          1         2  37.0              0   
6   d3b8ce26-9c89     500000    1          1         2  29.0              0   
7   d6e1911b-54d0     100000    2          2         2  23.0              0   
8   8421d3b3-2b54     140000    2          3         1  28.0              0   
9   62a2fd79-cd90      20000    1          3         2  35.0             -2   
10  b73c34fc-91cf     200000    2          3         2  34.0              0   
11  4777689a-5131     260000    2          1         2  51.0             -1   
12  b814ac5c-3d88     630000    2          2         2  41.0             -1   
13  a17b23d4-357b      70000    1          2         2  30.0              1   
14  0d6f99ad-84e9     250000    1          1         2  29.0              0   
15  8f023df0-a9ea      50000    2          3         3  23.0              1   
16  cbf648a6-98aa      20000    1          1         2  24.0              0   
17  f0ba5e02-0c0f     320000    1          1         1  49.0  Not available   
18  cb258690-8ea2     360000    2          1         1  49.0              1   
19  f6f0307a-b75f     180000    2          1         2  29.0              1   
20  47a4cdaa-aa20     130000    2          3         2  39.0              0   
21  834b2cac-750e     120000    2          2         1  39.0             -1   
22  c1d1d0f2-6aec      70000    2          2         2  26.0              2   
23  66a293d5-6384     450000    2          1         1  40.0             -2   
24  817396f7-aad1      90000    1          1         2  23.0              0   
25  275d9f10-f748      50000    1          3         2  23.0              0   
26  7784e5a8-a66c      60000    1          1         2  27.0              1   
27  99ccbfaa-715b      50000    2          3         2  30.0              0   
28  aecb449b-a999      50000    2          3         1  47.0  Not available   
29  57ea9a1f-cdd5      50000    1          1         2  26.0  Not available   
30  72c682a7-c8af     230000    2          1         2  27.0             -1   
31  a8eb0be8-23bb      50000    1          2         2  33.0              2   
32  3b76b111-03c1     100000    1          1         2  32.0              0   
33  4816a117-ae08     500000    2          2         1  54.0             -2   
34  e5a38e8f-e086     500000    1          1         1  58.0             -2   
35  ffbd7e58-a499     160000    1          1         2  30.0             -1   
36  d60f6ef1-01b2     280000    1          2         1  40.0              0   
37  db223ce8-182a      60000    2          2         2  22.0              0   
38  7f9a5e41-f41b      50000    1          1         2  25.0              1   
39  fce63a25-7f41     280000    1          1         2  31.0             -1   
40  713f9577-6c3b     360000    1          1         2  33.0              0   
41  2f10b695-b734      70000    2          1         2  25.0              0   
42  e59c0e10-21ff      10000    1          2         2  22.0              0   
43  9a066731-10e0     140000    2          2         1  37.0              0   
44  6a6245d9-e13b      40000    2          1         2  30.0              0   
45  cbea9a82-8ea3     210000    1          1         2  29.0             -2   
46  c479dcdd-9af9      20000    2          1         2  22.0              0   
47  99a3df27-dfb1     150000    2          5         2  46.0              0   
48  f4dbda9d-85da     380000    1          2         2  32.0             -1   
49  9ea8efa1-de6c      20000    1          1         2  24.0     

In [2]:
df_limpa.describe()

LIMIT_BAL           SEX     EDUCATION      MARRIAGE           AGE  \
count    29685.000000  29685.000000  29685.000000  29685.000000  29685.000000   
mean    167519.948796      1.603335      1.852788      1.552198     35.481354   
std     129716.478534      0.489214      0.789465      0.521869      9.212199   
min      10000.000000      1.000000      0.000000      0.000000     21.000000   
25%      50000.000000      1.000000      1.000000      1.000000     28.000000   
50%     140000.000000      2.000000      2.000000      2.000000     34.000000   
75%     240000.000000      2.000000      2.000000      2.000000     41.000000   
max    1000000.000000      2.000000      6.000000      3.000000     79.000000   

              PAY_2         PAY_3         PAY_4         PAY_5         PAY_6  \
count  29685.000000  29685.000000  29685.000000  29685.000000  29685.000000   
mean      -0.134277     -0.166077     -0.221627     -0.266700     -0.291090   
std        1.197440      1.197278      1.168278      1.133156      1.150653   
min       -2.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       -1.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        8.000000      8.000000      8.000000      8.000000      8.000000   

       ...      BILL_AMT4      BILL_AMT5      BILL_AMT6       PAY_AMT1  \
count  ...   29685.000000   29685.000000   29685.000000   29685.000000   
mean   ...   43245.439279   40307.628263   38888.682095    5672.886812   
std    ...   64276.899483   60787.204536   59588.111810   16616.452239   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2329.000000    1760.000000    1257.000000    1000.000000   
50%    ...   19040.000000   18106.000000   17067.000000    2100.000000   
75%    ...   54515.000000   50244.000000   49252.000000    5007.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

           PAY_AMT2       PAY_AMT3       PAY_AMT4       PAY_AMT5  \
count  2.968500e+04   29685.000000   29685.000000   29685.000000   
mean   5.917545e+03    5229.295570    4826.770827    4805.203840   
std    2.310628e+04   17650.360969   15607.256127   15311.798994   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.320000e+02     390.000000     296.000000     250.000000   
50%    2.008000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4508.000000    4015.000000    4041.000000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

            PAY_AMT6  default payment next month  
count   29685.000000                29685.000000  
mean     5219.023008                    0.221054  
std     17798.365183                    0.414964  
min         0.000000                    0.000000  
25%       115.000000                    0.000000  
50%      1500.000000                    0.000000  
75%      4000.000000                    0.000000  
max    528666.000000                    1.000000  

[8 rows x 23 columns]

#### 1) Saber quantas colunas os dados contêm. (Podem ser de características, resposta ou metadados).
#### 2) Quantas linhas (amostras).

In [3]:
df.shape
# Aqui se manteve para poder ser utilizado a fonte original.

(30000, 25)

**1. R:** Neste arquivo de dados existem 25 colunas.  
**2. R:** Neste arquivo de dados existem 30000 linhas. Portanto, foram coletadas 30000 amostras.

#### 3) Que tipos de características existem. Quais são categóricas e quais são numéricas? (Os valores das características categóricas pertencem a classes discretas como “sim”, “não” ou “talvez”. Normalmente as características numéricas pertencem a uma escala numérica contínua, como as quantias em dólares).

In [4]:
df_categorica = df_limpa.drop(['LIMIT_BAL','AGE', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6'], axis = 1)
display(df_categorica)
df_categorica.info()

ID  SEX  EDUCATION  MARRIAGE PAY_1  PAY_2  PAY_3  PAY_4  \
0      798fc410-45c1    2          2         1     2      2     -1     -1   
1      8a8c8f3b-8eb4    2          2         2    -1      2      0      0   
2      85698822-43f5    2          2         2     0      0      0      0   
3      0737c11b-be42    2          2         1     0      0      0      0   
4      3b7f77cc-dbc0    1          2         1    -1      0     -1      0   
...              ...  ...        ...       ...   ...    ...    ...    ...   
29995  ecff42d0-bdc6    1          3         1     0      0      0      0   
29996  99d1fa0e-222b    1          3         2    -1     -1     -1     -1   
29997  95cdd3e7-4f24    1          2         2     4      3      2     -1   
29998  00d03f02-04cd    1          3         1     1     -1      0      0   
29999  15d69f9f-5ad3    1          2         1     0      0      0      0   

       PAY_5  PAY_6  default payment next month  
0         -2     -2                           1  
1          0      2                           1  
2          0      0                           0  
3          0      0                           0  
4          0      0                           0  
...      ...    ...                         ...  
29995      0      0                           0  
29996      0      0                           0  
29997      0      0                           1  
29998      0     -1                           1  
29999      0      0                           1  

[29685 rows x 11 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 29685 entries, 0 to 29999
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          29685 non-null  object
 1   SEX                         29685 non-null  int64 
 2   EDUCATION                   29685 non-null  int64 
 3   MARRIAGE                    29685 non-null  int64 
 4   PAY_1                       29685 non-null  object
 5   PAY_2                       29685 non-null  int64 
 6   PAY_3                       29685 non-null  int64 
 7   PAY_4                       29685 non-null  int64 
 8   PAY_5                       29685 non-null  int64 
 9   PAY_6                       29685 non-null  int64 
 10  default payment next month  29685 non-null  int64 
dtypes: int64(9), object(2)
memory usage: 2.7+ MB


**3) R:** Características Categóricas: São no total 11 características categóricas, sendo elas:**  
**ID** - É um identificador único para as linhas com letras e números 
**SEX**, **Education**, **Marriage** - São características categóricas discretas, utilizam números para ditar características.
**PAY_1 à PAY_6** - Estas colunas não possuem valores contínuos como os de pagamento, mas, representam o índice de atraso de cada pagamento.
**default payment next month** - Através do número 0 e 1 representam respectivamente se o cliente é adimplente(0) ou inadimplente(1).

In [5]:
df_numerica = df_limpa.drop(['ID', 'SEX', 'EDUCATION', 'MARRIAGE', 'PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'default payment next month'], axis = 1)
display(df_numerica)

LIMIT_BAL   AGE  BILL_AMT1  BILL_AMT2  BILL_AMT3  BILL_AMT4  BILL_AMT5  \
0          20000  24.0       3913       3102        689          0          0   
1         120000  26.0       2682       1725       2682       3272       3455   
2          90000  34.0      29239      14027      13559      14331      14948   
3          50000  37.0      46990      48233      49291      28314      28959   
4          50000  57.0       8617       5670      35835      20940      19146   
...          ...   ...        ...        ...        ...        ...        ...   
29995     220000  39.0     188948     192815     208365      88004      31237   
29996     150000  43.0       1683       1828       3502       8979       5190   
29997      30000  37.0       3565       3356       2758      20878      20582   
29998      80000  41.0      -1645      78379      76304      52774      11855   
29999      50000  46.0      47929      48905      49764      36535      32428   

       BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  
0              0         0       689         0         0         0         0  
1           3261         0      1000      1000      1000         0      2000  
2          15549      1518      1500      1000      1000      1000      5000  
3          29547      2000      2019      1200      1100      1069      1000  
4          19131      2000     36681     10000      9000       689       679  
...          ...       ...       ...       ...       ...       ...       ...  
29995      15980      8500     20000      5003      3047      5000      1000  
29996          0      1837      3526      8998       129         0         0  
29997      19357         0         0     22000      4200      2000      3100  
29998      48944     85900      3409      1178      1926     52964      1804  
29999      15313      2078      1800      1430      1000      1000      1000  

[29685 rows x 14 columns]

**3) R:** Características Numéricas: São 14 colunas numéricas, pois apresentam valores de idade ou de dinheiro, sendo elas:   	LIMIT_BAL 	AGE 	BILL_AMT1 	BILL_AMT2 	BILL_AMT3 	BILL_AMT4 	BILL_AMT5 	E BILL_AMT6

#### 4) Qual é a aparência dos dados segundo essas características. (Para saber isso, você pode examinar o intervalo de valores em características numéricas ou a frequência de classes diferentes em características categóricas, por exemplo).

In [6]:
df_numerica.value_counts(df_limpa['AGE'], sort=False)

AGE
21.0      67
22.0     553
23.0     913
24.0    1113
25.0    1176
26.0    1243
27.0    1466
28.0    1394
29.0    1592
30.0    1388
31.0    1205
32.0    1144
33.0    1137
34.0    1152
35.0    1098
36.0    1095
37.0    1027
38.0     936
39.0     941
40.0     861
41.0     815
42.0     788
43.0     660
44.0     695
45.0     611
46.0     566
47.0     498
48.0     462
49.0     446
50.0     405
51.0     336
52.0     299
53.0     324
54.0     243
55.0     203
56.0     175
57.0     121
58.0     121
59.0      82
60.0      65
61.0      56
62.0      42
63.0      31
64.0      31
65.0      23
66.0      25
67.0      16
68.0       5
69.0      15
70.0      10
71.0       3
72.0       3
73.0       4
74.0       1
75.0       3
79.0       1
Name: count, dtype: int64

In [7]:
# Aqui foi identificado que as linhas onde a coluna AGE é igual à 0 não há valores, somente a linha de crédito, por conta disso
# Todo o código foi adaptado para trabalhar encima da tabela sem estas linhas.
# df['AGE'] = df['AGE'].replace(0, np.nan)
# df_limpa = df.dropna() 


In [8]:
df_numerica.describe()

LIMIT_BAL           AGE      BILL_AMT1      BILL_AMT2  \
count    29685.000000  29685.000000   29685.000000   29685.000000   
mean    167519.948796     35.481354   51184.178103   49140.322553   
std     129716.478534      9.212199   73578.295929   71091.014997   
min      10000.000000     21.000000 -165580.000000  -69777.000000   
25%      50000.000000     28.000000    3560.000000    2984.000000   
50%     140000.000000     34.000000   22381.000000   21197.000000   
75%     240000.000000     41.000000   67028.000000   63992.000000   
max    1000000.000000     79.000000  964511.000000  983931.000000   

          BILL_AMT3      BILL_AMT4      BILL_AMT5      BILL_AMT6  \
count  2.968500e+04   29685.000000   29685.000000   29685.000000   
mean   4.699076e+04   43245.439279   40307.628263   38888.682095   
std    6.930111e+04   64276.899483   60787.204536   59588.111810   
min   -1.572640e+05 -170000.000000  -81334.000000 -339603.000000   
25%    2.663000e+03    2329.000000    1760.000000    1257.000000   
50%    2.010700e+04   19040.000000   18106.000000   17067.000000   
75%    6.016700e+04   54515.000000   50244.000000   49252.000000   
max    1.664089e+06  891586.000000  927171.000000  961664.000000   

            PAY_AMT1      PAY_AMT2       PAY_AMT3       PAY_AMT4  \
count   29685.000000  2.968500e+04   29685.000000   29685.000000   
mean     5672.886812  5.917545e+03    5229.295570    4826.770827   
std     16616.452239  2.310628e+04   17650.360969   15607.256127   
min         0.000000  0.000000e+00       0.000000       0.000000   
25%      1000.000000  8.320000e+02     390.000000     296.000000   
50%      2100.000000  2.008000e+03    1800.000000    1500.000000   
75%      5007.000000  5.000000e+03    4508.000000    4015.000000   
max    873552.000000  1.684259e+06  896040.000000  621000.000000   

            PAY_AMT5       PAY_AMT6  
count   29685.000000   29685.000000  
mean     4805.203840    5219.023008  
std     15311.798994   17798.365183  
min         0.000000       0.000000  
25%       250.000000     115.000000  
50%      1500.000000    1500.000000  
75%      4041.000000    4000.000000  
max    426529.000000  528666.000000

**Analisando as classes numéricas, é possível identificar que a empressa possui clientes com idade entre 21 e 79 anos com média de 35 anos. Iremos nos aprofundar mais para responder a questão 4**

In [9]:
df_analise = df_limpa[['AGE','default payment next month']].copy()
bins = [21, 28, 34, 41, 79]
# As faixas etárias foram decididas de acordo com os valores descritos pelo .describe.
df_analise['AGE'] = pd.cut(df_analise['AGE'], bins, right=False)
df_analise.value_counts(df_analise['AGE'], sort=False)

AGE
[21, 28)    6531
[28, 34)    7860
[34, 41)    7110
[41, 79)    8183
Name: count, dtype: int64

In [10]:
df_analise.value_counts(sort=False)

AGE       default payment next month
[21, 28)  0                             4950
          1                             1581
[28, 34)  0                             6332
          1                             1528
[34, 41)  0                             5614
          1                             1496
[41, 79)  0                             6226
          1                             1957
Name: count, dtype: int64

In [11]:
df_analise.value_counts(sort=False, normalize=True)

AGE       default payment next month
[21, 28)  0                             0.166757
          1                             0.053261
[28, 34)  0                             0.213314
          1                             0.051476
[34, 41)  0                             0.189125
          1                             0.050398
[41, 79)  0                             0.209743
          1                             0.065928
Name: proportion, dtype: float64

**4) R: Analisando os dados através do value counts podemos ver que:**
**A quantidade de clientes em cada faixa etária apresentam os seguintes dados:**  
- 21-27 anos com 6531 clientes (menor faixa etária)
- 28-33 anos com 7860 clientes
- 34-40 anos com 7110 clientes
- 41-79 anos com 8183 clientes (maior faixa etária)  
Levando em consideração as taxas de inadimplentes/adimplentes com o total de clientes, podemos ver a seguinte ordem:    
**Inadimplentes:**  
- 21-27(5,32%/1581)
- 28-33(5,14%/1528)
- 34-40(5,03%/1496) (menor taxa de inadimplentes)
- 41-79(6,59%/1957) (maior taxa de inadimplentes)    
**Adimplentes:**
- 21-27(16,67%/4950) (menor taxa de adimplentes)
- 28-33(21,33%/6332) (maior taxa de adimplentes)
- 34-40(18,91%/5614)
- 41-79(20,97%/6226)  
**Diferença entre Adimplentes e Inadimplentes**    
- 21-27 - 11,35%/3.369 positivos 
- 28-33 - 16,19%/4.804 positivos
- 34-40 - 13,88%/4.118 positivos
- 41-79 - 15,38%/4.269 positivos  
**Agora iremos analisar os números individualmente**    
- Os clientes de **21 a 27 anos** representam a **menor parte de clientes**, mas, também possuem um **índice de inadimplência equiparado** com as outras faixas etárias, representado a parte com **menor estabilidade financeira** e com **maior risco de inadimplência**.
- Os clientes de **28 a 33 anos** apresentam a **maior diferença positiva**, representando a faixa etária com **maior estabilidade financeira**, assim possuindo um risco muito baixo de inadimplência.
- Os clientes de **34 a 40 anos** possuem a **menor taxa de inadimplentes** em ralação ao total, mas, possuem uma **taxa de adimplentes relativamente menor** que as outras faixas etárias, em relação aos valores positivos, essa faixa etária mantêm uma boa relação entre adimplentes e inadimplentes, mostrando que sua **estabilidade financeira é positiva** levando a possuir um risco baixo de inadimplência.
- Os clientes de **41 79 anos** **são a maioria da base de clientes e ocupam o 2º lugar de maior taxa de adimplentes**, porém, possuem uma **elevada taxa de clientes inadimplentes**, considerando que pode ser um resultado da quantidade de clientes presentes nela, mas, considerando isso, esta faixa etária **possui um risco médio**, mas, ainda assim mantêm uma **boa estabilidade financeira**. O seu risco médio leva em consideração que a **faixa etária anterior (34 a 40 anos) possui 1000 clientes a menos** e apresentam um **número positivo próximo a esta faixa etária.**

**5)Existem dados faltando?**

In [12]:
(df_limpa == "Not available").sum()

ID                               0
LIMIT_BAL                        0
SEX                              0
EDUCATION                        0
MARRIAGE                         0
AGE                              0
PAY_1                         3021
PAY_2                            0
PAY_3                            0
PAY_4                            0
PAY_5                            0
PAY_6                            0
BILL_AMT1                        0
BILL_AMT2                        0
BILL_AMT3                        0
BILL_AMT4                        0
BILL_AMT5                        0
BILL_AMT6                        0
PAY_AMT1                         0
PAY_AMT2                         0
PAY_AMT3                         0
PAY_AMT4                         0
PAY_AMT5                         0
PAY_AMT6                         0
default payment next month       0
dtype: int64

**5)R:** Sim, através da análise do banco de dados, foi possível identificar 2 tipos de dados faltantes:
- Linhas de Crédito sem dados: cerca de 315 linhas no banco de dados apresentam ID, mas deixam de lado todas as outras colunas, ficando cada uma com o valor 0. Considerando que estas linhas não possuem dado nenhum que possa contribuir com a análise, elas foram consideradas como dados faltantes.
- Coluna "PAY_1" como "Not available": Foram identificadas cerca de 3021 linhas onde a coluna "PAY_1" apresentavam a palavra "Not available" que simplesmente significa "Indisponível", sendo um termo utilizado neste banco de dados para representar a falta de valores.

Assim podemos considerar que 3336 linhas do banco de dados possuem um ou mais dados faltando.